service/debugger.go

```diff
@@ -0,0 +1,402 @@
+package service
+
+import (
+	"fmt"
+	"regexp"
+	"runtime"
+
+	"github.com/golang/glog"
+	sys "golang.org/x/sys/unix"
+
+	"github.com/derekparker/delve/proctl"
+	"github.com/derekparker/delve/service/api"
+)
+
+// Debugger provides a thread-safe DebuggedProcess service. Instances of
+// Debugger can be exposed by other services.
+type Debugger struct {
+	processFuncs chan func(*proctl.DebuggedProcess) error
+	stop         chan bool
+}
+
+func NewDebugger(stop chan bool) *Debugger {
+	debugger := &Debugger{
+		processFuncs: make(chan func(*proctl.DebuggedProcess) error),
+		stop:         stop,
+	}
+	return debugger
+}
+
+// withProcess facilitates thread-safe access to the DebuggedProcess. All
+// interaction with DebuggedProcess should occur via calls to withProcess, and
+// the functions placed on the processFuncs channel should be consumed and
+// executed from the same thread as the DebuggedProcess.
+//
+// This is convenient because it allows things like HTTP handlers in
+// goroutines to work with the DebuggedProcess with synchronous semantics.
+func (d *Debugger) withProcess(f func(*proctl.DebuggedProcess) error) error {
+	done := make(chan interface{})
```

Comment:
Why not just `make(chan error)` and below `done <- f(proc)` and below that: `return <-done`.